### PR TITLE
API method 'getJob', index jobs (needs tests)

### PR DESCRIPTION
**Pull request by ricardofbarros**
Continuing from issue #54

I'm missing the tests but see if this make any sense, basically now you have a new API method `getJob`, which can be used internally to resolve and get the job and developers can use it as well when they just know the name of the job but not the index, for e.g.

```js
var schedule = require('node-schedule');

schedule.scheduleJob('test', '42 * * * *', function() {
    console.log('The answer to life, the universe, and everything!');
});

var jobTest = schedule.getJob('test');
```

So give me any feedback if this is the right path. If it is I'll make the tests for this, btw using `getJob`internally the current tests all have passed, so if there is tests to be done it will be for the new method `getJob` and the dilemma in the line 461.

### CODE REVIEW

**Comment by tejasmanohar** [comment]
I like it. Would def have to be a major release. Want to hear feedback from @santigimeno @jonhester

File: lib/schedule.js [hunk 1]
```diff
@@ -438,16 +443,49 @@ function cancelJob(job) {
...
       }
     }
-  } else if (typeof job == 'string' || job instanceof String) {
-    if (job in scheduledJobs && scheduledJobs.hasOwnProperty(job)) {
-      success = scheduledJobs[job].cancel();
-      if (success) scheduledJobs[job] = null;
-    }
+  } else {
+    var findJob = getJob(job);
+    if (findJob instanceof Job) return cancelJob(findJob);
```
Comment (santigimeno) [inline lib/schedule.js]: style: use braces and move statement to new line

**Comment by santigimeno** [comment]
Looks great :+1: . After adding some tests I think it should fine to merge. Yes, it should be a major version.
One thing about the API: after this change we'll have the `scheduledJobs` property to retrieve all jobs and the `getJob` function to retrieve a specific job. It would be nice to have some consistency there.

**Comment by ricardofbarros** [comment]
Okay,
so you're saying to add a new function `getJobs` an alias to `scheduledJobs`?

**Comment by santigimeno** [comment]
Yeah, that could work. Let's see what others think

**Comment by jonhester** [comment]
Looks good to me. This will be super useful. Definitely want to have some tests, especially since we don't currently have full coverage.

**Comment by ricardofbarros** [comment]
Okay, I'll start looking how the tests are organized and add coverage to this changes.

**Comment by tejasmanohar** [comment]
Yup. With tests, I'm ready to merge. Good work :+1:

**Comment by DaAwesomeP** [comment]
If I were to input a number instead of a name, would that work? I guess a number in a string would always work. An ID system is most efficient for automated scheduling.

**Comment by ricardofbarros** [comment]
A number on `getJob`? Yes it would work

**Comment by DaAwesomeP** [comment]
Never mind I just looked at the commit. Is it possible to get the ID of a job when creating it?

**Comment by ricardofbarros** [comment]
Good point maybe we should add a key `id` to the job object, so we know inside the object what id it has. Another situation I remebered  is that when I for e.g. cancel one job and try add 2 more. Something bad will happen from that situation, the last schedule added will overwrite the last one I need to make ids incremental.

I will explain with a code example

```js

var schedule = require('node-schedule');

schedule.scheduleJob('test', '42 * * * *', function() {
    console.log('The answer to life, the universe, and everything!');
});

schedule.scheduleJob( '5 * * * *', function() {
    console.log('ring ring!');
});

/* So now our scheduledJobs should be something
* like this:
*    '0': {
*      ...
* }, 
* '1': {
*     ...
* } 
*/

```

File: lib/schedule.js (hunk 1 above)
Comment (jonhester) [inline lib/schedule.js]: @santigimeno want me to change the linter to fail on this?

**Comment by santigimeno** [comment]
+1

**Comment by tejasmanohar** [comment]
+1 @DaAwesomeP nice catch.

File: lib/schedule.js (hunk 1 above)
Comment (tejasmanohar) [inline lib/schedule.js]: @jonhester yeah

**Comment by DaAwesomeP** [comment]
Not only do the IDs need to be incremental, but they need to not take the place of a cancelled job. For example, If I create jobs 1,2, and 3, and then I delete jobs 1 and 3, the next job should be 4, not 1 or 3. A simply counter variable should do. This is the alternative to getting the last object's ID.

The variable should be the value of the next job to be created, not what the last job created was. Otherwise they will start 1 instead of 0. So, when the counter variable is defined, it should equal 0.

**Comment by DaAwesomeP** [comment]
Lastly, make sure that you can modify a job as well as cancel. This would simply grab a job by the ID and then overwrite it (or delete and re-add it). Ideally you would have the option to change the Date/Cron, callback function, or both at once.

**Comment by DaAwesomeP** [comment]
Any progress?

**Comment by tejasmanohar** [comment]
cc @ricardofbarros

**Comment by ricardofbarros** [comment]
Okay added some tests, take a look.

File: test/convenience-method-test.js [hunk 2]
```diff
@@ -17,6 +17,49 @@ module.exports = {
 
       job.cancel();
       test.done();
+    },
+    "Job ids are incremental": function(test) {
+      // Create 4 new jobs
+      var job1 = schedule.scheduleJob(new Date(Date.now() + 1000), function() {});
```
Comment (santigimeno) [inline test/convenience-method-test.js]: test.expect(5);

**Comment by santigimeno** [comment]
It looks good to me. Maybe, as this is going to be in a major version, remove `scheduledJobs` as you already comment in the code

**Comment by ricardofbarros** [comment]
Added the expect(5). Sure if it's going to be a major release why not remove it.

**Comment by santigimeno** [comment]
Also, rebase to current master

**Comment by ricardofbarros** [comment]
Why rebase and not merge?

**Comment by santigimeno** [comment]
I mean rebase you branch to current master to see if there are conflicts

**Comment by santigimeno** [comment]
So we can merge later without problems

**Comment by ricardofbarros** [comment]
It had just a small conflict on cancelJob

**Comment by tejasmanohar** [comment]
@jonhester Is there an issue w/ Travis? Looks like it's still building in-progress for hours now.

**Comment by DaAwesomeP** [comment]
Look here: https://travis-ci.org/node-schedule/node-schedule/builds/55517024
GitHub simply hasn't updated its status. It's happened to me before on other pull requests too (its a GitHub bug).

**Comment by jonhester** [comment]
yeah, still annoying.

**Comment by tejasmanohar** [comment]
Alright, so looks like build fails on 0.10 due to an incompatibility w/ coveralls module but not an issue w/ @ricardofbarros code. Am I right?

**Comment by jonhester** [comment]
Exactly. Looks like coveralls may have been down briefly. All the years passed.

**Comment by tejasmanohar** [comment]
Alright. How do you guys think we should handle merging this? What do you think about a 1.0 release branch that we merge in stable updates that are going to be released in 1.0 before they're ready for master? Open to advice on how to handle this / what's the convention.

**Comment by jonhester** [comment]
That's my vote. It will make it a lot easier to start refactoring some things to have a branch to merge against. And then when we are ready we just merge into master and publish 1.0.

**Comment by DaAwesomeP** [comment]
Does this have breaking API changes? If not then it should technically go into the "minor" stage of subversion system.

**Comment by tejasmanohar** [comment]
@DaAwesomeP API structure changed yeah, they're indexed jobs now.

**Comment by tejasmanohar** [comment]
Let's follow https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow and set the below as the **etiquette** for outside contributions.

https://www.atlassian.com/git/tutorials/comparing-workflows/forking-workflow/

@jonhester see #129 since I'd want to make sure that the development branches you're currently maintaining on this repo are properly setup on origin before moving on.

**Comment by tejasmanohar** [comment]
After implement git flow, you can go ahead and file another PR looking at @ricardofbarros's fork (so his work will still be credited) but instead of merging into the branch master, into the 1.0 release. How does that sound? (to everyone)

**Comment by tejasmanohar** [comment]
(you can still comment here, we'll have to open a new PR either way when new branch is setup)

**Comment by DaAwesomeP** [comment]
We should probably fix the test so that the big pull will be recorded as passing.

**Comment by jonhester** [comment]
@tejasmanohar New develop branch is setup, so @ricardofbarros can go ahead a submit a pull request against that branch.

**Comment by DaAwesomeP** [comment]
This can be installed via NPM:
```bash
npm install git://github.com/node-schedule/node-schedule#develop
```

**Pull request closed without merging**